e2e/nomostest/reset.go

```diff
@@ -0,0 +1,513 @@
+		// dependencies that have owners references. Otherwise the reconciler
+		// and dependenencies will be garbage collected before the finalizer
+		// can delete the managed resources.
+		// TODO: Remove explicit Background policy after the reconciler-manager finalizer is added.
+		nt.T.Logf("[RESET] Deleting RootSync %s", rsNN)
+		if err := nt.Delete(rs, client.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil {
+			return err
+		}
+	}
+	tg := taskgroup.New()
+	for _, item := range rsList {
+		rs := &item
+		rsNN := client.ObjectKeyFromObject(rs)
+		nt.T.Logf("[RESET] Waiting for deletion of RootSync %s ...", rsNN)
+		tg.Go(func() error {
+			return WatchForNotFound(nt, kinds.RootSyncV1Beta1(), rs.Name, rs.Namespace)
+		})
+	}
+	return tg.Wait()
+}
+
+// ResetRepoSyncs cleans up one or more RepoSyncs and all their managed objects.
+// Use this for cleaning up RepoSyncs in tests that use delegated control.
+//
+// To ensure the reconcile finalizer has permission to delete managed resources,
+// ClusterRole and RoleBindings will be created and then later deleted.
+// This also cleans up any CRs, RBs, and CRBs left behind by delegated control.
+func ResetRepoSyncs(nt *NT, rsList []v1beta1.RepoSync) error {
+	if len(rsList) == 0 {
+		// Clean up after `setupDelegatedControl`
+		return deleteRepoSyncClusterRole(nt)
+	}
+
+	nt.T.Logf("[RESET] Deleting RepoSyncs (%d)", len(rsList))
+
+	record := newResetRecord()
+
+	// Apply ClusterRole with the permissions specified by this test.
+	rsCR := nt.RepoSyncClusterRole()
+	if err := nt.Apply(rsCR); err != nil {
+		return err
+	}
+
+	for _, item := range rsList {
+		rs := &item
+		rsNN := client.ObjectKeyFromObject(rs)
+		record.Objects[rsNN] = struct{}{}
+		record.ObjectNamespaces[rsNN.Namespace] = struct{}{}
+
+		// If managed, skip direct deletion
+		if manager, found := rs.GetAnnotations()[string(metadata.ResourceManagerKey)]; found {
+			record.ManagedObjects[rsNN] = struct{}{}
+			nt.T.Logf("[RESET] RepoSync %s managed by %q", rsNN, manager)
+			if !IsDeletionPropagationEnabled(rs) {
+				// If you go this error, make sure your test cleanup ensures
+				// that the managed RepoSync has deletion propagation enabled.
+				return errors.Errorf("RepoSync %s managed by %q does NOT have deletion propagation enabled: test reset incomplete", rsNN, manager)
+			}
+			continue
+		}
+
+		// Enable deletion propagation, if not enabled
+		if EnableDeletionPropagation(rs) {
+			nt.T.Logf("[RESET] Enabling deletion propagation on RepoSync %s", rsNN)
+			if err := nt.Update(rs); err != nil {
+				return err
+			}
+			if err := WatchObject(nt, kinds.RepoSyncV1Beta1(), rs.Name, rs.Namespace, []Predicate{
+				HasFinalizer(metadata.ReconcilerFinalizer),
+			}); err != nil {
+				return err
+			}
+		}
+
+		// Grant the reconcile the permissions specified by this test.
+		rsCRB := RepoSyncRoleBinding(rsNN)
+		if err := nt.Apply(rsCRB); err != nil {
```

Comment:
Maybe instead of having the for loops below to clean these up, we could add a `defer` statement here to delete the `RoleBinding`. WDYT?

Reply:
Defer could work. Or I could build up the Delete & Wait task group now and defer the tg.Wait()...

But do we want to defer in Reset/Clean (see other comment)?